jkube-kit/enricher/generic/src/main/java/org/eclipse/jkube/enricher/generic/AbstractLabelEnricher.java

```diff
@@ -133,14 +145,47 @@ public void enrich(PlatformMode platformMode, KubernetesListBuilder builder) {
       @Override
       public void visit(ObjectMetaBuilder element) {
         Map<String, String> labels = Optional.ofNullable(element.getLabels()).orElse(new HashMap<>());
-        MapUtil.mergeIfAbsent(labels, createLabels(false));
+        MapUtil.mergeIfAbsent(labels, createLabels(false, Collections.emptyMap()));
         element.withLabels(labels);
       }
     });
   }
 
-  private Map<String, String> mergedSelectors(Map<String, String> originalSelectors) {
-    MapUtil.mergeIfAbsent(originalSelectors, createLabels(true));
+  protected Map<String, String> extractLabelsConfiguredViaResourceConfig(List<Supplier<Properties>> labelsConfiguredViaResourceConfigSupplierList) {
+    Map<String, String> labelsFromResourceConfig = new HashMap<>();
+    List<Supplier<Properties>> labelSources = new ArrayList<>();
+    labelSources.add(() -> getResourceConfigLabels().getAll());
+    labelSources.addAll(labelsConfiguredViaResourceConfigSupplierList);
+    for (Supplier<Properties> labelSupplier : labelSources) {
+      Map<String, String> labels = convertLabelsToMap(labelSupplier);
+      labelsFromResourceConfig.putAll(labels);
+    }
+    return labelsFromResourceConfig;
+  }
+
+  private Map<String, String> convertLabelsToMap(Supplier<Properties> labelSupplier) {
+    Properties labelsAsProps = labelSupplier.get();
+    Map<String, String> labels = new HashMap<>();
+    if (labelsAsProps != null) {
+      for (Map.Entry<Object, Object> e : labelsAsProps.entrySet()) {
+        String key = (String) e.getKey();
+        String value = (String) e.getValue();
+        labels.put(key, value);
+      }
+    }
+    return labels;
+  }
+
+  private Map<String, String> mergedSelectors(Map<String, String> originalSelectors, List<Supplier<Properties>> labelSupplierList) {
+    Map<String, String> labelsFromResourceConfig = extractLabelsConfiguredViaResourceConfig(labelSupplierList);
+    MapUtil.mergeIfAbsent(originalSelectors, createLabels(true, labelsFromResourceConfig));
     return originalSelectors;
   }
```

Comment:
Have you checked PropertiesUtil and MapUtil to see if some of the provided methods can be reused?

I'm kind of sure we can easily remove the convertLabelsToMap method and replace it with toMap(labelSupplier.get()), but we can very likely simplify more

Reply:
mergeMaps + toMap should really allow you to remove extractLabelsConfiguredViaResourceConfig + convertLabelsToMap and make it more clear in the calling methods.

I'm not sure if we're being clear about precedence either.

Also MapUtil.mergeMaps can be changed to `MapUtil.mergeMaps(Map<K, V> maps...)` to be able to support a variable number of maps